src/sunadjointcheckpointscheme/fixed/sunadjointcheckpointscheme_fixed.c

```diff
@@ -0,0 +1,285 @@
+    else if (errcode == SUN_ERR_DATANODE_NODENOTFOUND)
+    {
+      step_data_node = NULL;
+    }
+    else { SUNCheckCall(errcode); }
+  }
+  else { step_data_node = IMPL_MEMBER(self, current_load_step_node); }
+
+  if (!step_data_node)
+  {
+    SUNLogExtraDebug(SUNCTX_->logger, "step-not-found",
+                     "step_num = %d, stage_num = %d", step_num, stage_num);
+    return SUN_ERR_CHECKPOINT_NOT_FOUND;
+  }
+
+  SUNLogExtraDebug(SUNCTX_->logger, "step-loaded",
+                   "step_num = %d, stage_num = %d", step_num, stage_num);
+
+  SUNDataNode solution_node = NULL;
+  if (IMPL_MEMBER(self, keep) || peek)
+  {
+    SUNLogExtraDebug(SUNCTX_->logger, "try-load-stage",
+                     "keep = 1, step_num = %d, stage_num = %d", step_num,
+                     stage_num);
+    errcode = SUNDataNode_GetChild(step_data_node, stage_num, &solution_node);
+    if (errcode == SUN_ERR_DATANODE_NODENOTFOUND) { solution_node = NULL; }
+    else { SUNCheckCall(errcode); }
+  }
+  else
+  {
+    sunbooleantype has_children = SUNFALSE;
+    SUNCheckCall(SUNDataNode_HasChildren(step_data_node, &has_children));
+
+    if (has_children)
+    {
+      SUNLogExtraDebug(SUNCTX_->logger, "try-load-stage",
+                       "keep = 0, step_num = %d, stage_num = %d", step_num,
+                       stage_num);
+      errcode = SUNDataNode_RemoveChild(step_data_node, stage_num,
+                                        &solution_node);
+      if (errcode == SUN_ERR_DATANODE_NODENOTFOUND) { solution_node = NULL; }
+      else { SUNCheckCall(errcode); }
+    }
+
+    SUNCheckCall(SUNDataNode_HasChildren(step_data_node, &has_children));
+    if (!has_children)
+    {
+      char* key = sunSignedToString(step_num);
+      SUNLogExtraDebug(SUNCTX_->logger, "remove-step", "step_num = %d", step_num);
+      SUNCheckCall(SUNDataNode_RemoveNamedChild(IMPL_MEMBER(self, root_node),
+                                                key, &step_data_node));
+      free(key);
+      SUNCheckCall(SUNDataNode_Destroy(&step_data_node));
+    }
+  }
+
+  if (!solution_node)
+  {
+    SUNLogExtraDebug(SUNCTX_->logger, "stage-not-found",
+                     "step_num = %d, stage_num = %d", step_num, stage_num);
+    return SUN_ERR_CHECKPOINT_NOT_FOUND;
+  }
+
+  SUNCheckCall(SUNDataNode_GetDataNvector(solution_node, *yout, t));
+  SUNLogExtraDebug(SUNCTX_->logger, "stage-loaded",
+                   "step_num = %d, stage_num = %d, t = %g", step_num, stage_num,
+                   *t);
+
+  /* Cleanup the checkpoint memory if need be */
+  if (!(IMPL_MEMBER(self, keep) || peek))
+  {
+    SUNCheckCall(SUNDataNode_Destroy(&solution_node));
+  }
+
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNAdjointCheckpointScheme_Destroy_Fixed(
+  SUNAdjointCheckpointScheme* self_ptr)
+{
+  SUNFunctionBegin((*self_ptr)->sunctx);
+
+  SUNAdjointCheckpointScheme self = *self_ptr;
+
+  SUNCheckCall(SUNDataNode_Destroy(&IMPL_MEMBER(self, root_node)));
+
+  free(GET_CONTENT(self));
```

Comment:
I think this is slightly more clear in this case
```suggestion
  free(self->content);
```